gordon/router.py

```diff
@@ -142,13 +172,22 @@ def _get_next_phase(self, event_msg):
                 msg = f'Adding message "{event_msg}" to success channel.'
                 event_msg.append_to_history(msg, event_msg.phase)
                 await self.success_channel.put(event_msg)
+            elif old_phase == 'publish':
+                # if we're here, can assume message successfully went thru
+                # the entire phase route
+                await self.metrics.incr('router-message-completed')
 
         except Exception as e:
             msg = f'Dropping message "{event_msg}" due to exception: "{e}"'
             event_msg.append_to_history(msg, event_msg.phase)
             logging.warn(msg, exc_info=e)
+            context = {'error': e.__class__.__name__}
+            await self.metrics.incr('router-message-dropped', context=context)
             await self._route(event_msg, 'cleanup')
 
+        finally:
+            await self._update_messages_in_flight(event_msg)
```

Comment:
you're calling `await self._update_messages_in_flight(event_msg)` twice in this function. is that needed?

Reply:
Yes it was intentional. The first time ensures that the message is in memory (if we've already seen it, then it does nothing except send the gauge metric, which is fine). And this second time to ensure that if a message is dropped or done, it gets removed from memory. All to ensure we give an accurate count of messages in flight.

Perhaps there's a better way to do this?

Reply:
It seems like there's two things in `_update_messages_in_flight`. Maybe could split out an ` _add_to_messages_in_flight` function?